lib/inventory/controller.go

```diff
@@ -418,6 +457,31 @@ func (c *Controller) handleControlStream(handle *upstreamHandle) {
 	}
 }
 
+func (c *Controller) timeReconciliation(handle *upstreamHandle, now time.Time) error {
+	id := rand.Uint64()
+	if err := handle.Send(c.closeContext, proto.DownstreamInventoryPing{ID: id}); err != nil {
+		return trace.Wrap(err)
+	}
+	rspC := make(chan pingResponse, 1)
+	handle.pings[id] = pendingPing{
+		start: now,
+		rspC:  rspC,
+	}
+	tracker := &handle.stateTracker
+	go func() {
+		select {
+		case <-c.closeContext.Done():
+			return
+		case pong := <-rspC:
+			tracker.mu.Lock()
+			tracker.systemClockDifference = c.clock.Since(pong.systemClock) - pong.reqDuration/2
```

Comment:
What's the point of checking the time difference again here? Don't we have the ping start time and the "duration" already?

Reply:
not sure if I understand correctly, do we need to store in state tracker time when request is send, response clock and request duration?

because when we request `timeReconciliation` from the `handleControlStream` loop, we make ping request to the handler receive system clock with some response delay, then store the time difference in state tracker and then in next heartbeat send this duration as part of `types.Instance`

Reply:
We should measure our local time as close to the receive as possible, so I think it would be better to save local time, remote time (if provided) and duration in the `pingResponse` object.

Reply:
@espadolini could you please take another look, if I do right inventory collection iteration. I've also just replaced in proto to expose clock data instead time difference duration